doc/monads/readers.lean

```diff
@@ -0,0 +1,196 @@
+statement in Lean can only be used inside a `do` block and is closely related to `pure`, but a
+little different. First the similarity is that `return` and `pure` both lift a pure value up to the
+Monad type. But `return` is a keyword so you do not need to parenthesize the expression like you do
+when using `pure`.  (Note: you can avoid parentheses when using `pure` by using the `<|` operator
+like we did above in the initial `getEnvDefault` function).  Furthermore, `return` can also cause an
+early `return` in a monadic function similar to how it can in an imperative language while `pure`
+cannot.
+
+> So technically if `return` is the last statement in a function it could be replaced with `pure <|`,
+but one could argue that `return` is still a little easier for most folks to read, just so long as
+you understand that `return` is doing more than other languages, it is also wrapping pure values in
+the monadic container type.
+
+> **Side note 2**: If the function `readerFunc3` also took some explicit arguments then you would have
+to write `(readerFunc3 args).run env` and this is a bit ugly, so Lean provides an infix operator
+`|>` that eliminiates those parens so you can write `readerFunc3 args |>.run env` and then you can
+chain multiple monadic actions like this `m1 args1 |>.run args2 |>.run args3` and this is the
+recommended style.  You will see this patten used heavily in Lean code.
+
+The `let env ← read` expression in `readerFunc1` unwraps the environment from the `ReaderM` so we
+can use it. Each type of monad might provide one or more extra functions like this, functions that
+become available only when you are in the context of that monad.
+
+Here the `readerFunc2` function uses the `bind` operator `>>=` just to show you that there are bind
+operations happening here.  The `readerFunc3` function uses the  `do` notation you learned about in
+[Monads](monads.lean.md) which hides that bind operation and can make the code look cleaner.
+
+The `do` notation with `let x ← readerFunc2` is also calling the `run` function to actually execute
+the monadic action, then it unwraps the value x, so it is really doing `let x ← readerFunc2.run`.
+
+The important difference here to the earlier code is that `readerFunc3` and `readerFunc2` no longer
+have an **explicit** Environment input parameter that needs to be passed along all the way to
+`readerFunc1`.  Instead, the `ReaderM` monad is taking care of that for you, which gives you the
+illusion of something like global state where the state is now available to all functions that use
+the `ReaderM` monad.
+
+The above code also introduces an important idea. Whenever you learn about a monad "X", there's
+often (but not always) a `run` function to execute that monad, and sometimes some additional
+functions like `read` that interact with the monad state.
+
+You might be wondering, how does the state actually move through the `ReaderM` monad? How can you
```

Comment:
"state" => "context"

Reply:
Fixed in https://github.com/leanprover/lean4/pull/1505/commits/6fed17209a17e523b02d2dd6b0cc773ec551738f